src/musig.rs

```diff
@@ -0,0 +1,1286 @@
+/// output of this function cannot be used to sign for any other public key.
+/// * `msg`: Optional [`Message`] that will be signed later on. Provide this for maximal misuse resistance.
+/// * `extra_rand`: Additional randomness for mis-use resistance. Provide this for maximal misuse resistance
+///
+/// Remember that nonce reuse will immediately leak the secret key!
+///
+/// # Errors:
+///
+/// * `ZeroSession`: if the `session_id` is supplied is all zeros.
+///
+/// Example:
+///
+/// ```rust
+/// # # [cfg(any(test, feature = "rand-std"))] {
+/// # use secp256k1_zkp::rand::{thread_rng, RngCore};
+/// # use secp256k1_zkp::{PublicKey, Secp256k1, SecretKey, new_musig_nonce_pair, MusigSessionId};
+/// # let secp = Secp256k1::new();
+/// // The session id must be sampled at random. Read documentation for more details.
+/// let session_id = MusigSessionId::new(&mut thread_rng());
+/// let sk = SecretKey::new(&mut thread_rng());
+/// let pk = PublicKey::from_secret_key(&secp, &sk);
+///
+/// // Supply extra auxillary randomness to prevent misuse(for example, time of day)
+/// let extra_rand : Option<[u8; 32]> = None;
+///
+/// let (_sec_nonce, _pub_nonce) = new_musig_nonce_pair(&secp, session_id, None, Some(sk), pk, None, None)
+///     .expect("non zero session id");
+/// # }
+/// ```
+pub fn new_musig_nonce_pair<C: Signing>(
+    secp: &Secp256k1<C>,
+    session_id: MusigSessionId,
+    key_agg_cache: Option<&MusigKeyAggCache>,
+    sec_key: Option<SecretKey>,
+    pub_key: PublicKey,
+    msg: Option<Message>,
+    extra_rand: Option<[u8; 32]>,
+) -> Result<(MusigSecNonce, MusigPubNonce), MusigNonceGenError> {
+    let cx = secp.ctx().as_ptr();
+    let extra_ptr = extra_rand
+        .as_ref()
+        .map(|e| e.as_ptr())
+        .unwrap_or(core::ptr::null());
+    let sk_ptr = sec_key
+        .as_ref()
+        .map(|e| e.as_c_ptr())
+        .unwrap_or(core::ptr::null());
+    let msg_ptr = msg
+        .as_ref()
+        .map(|ref e| e.as_c_ptr())
+        .unwrap_or(core::ptr::null());
+    let cache_ptr = key_agg_cache
+        .map(|e| e.as_ptr())
+        .unwrap_or(core::ptr::null());
+    unsafe {
+        let mut sec_nonce = MusigSecNonce(ffi::MusigSecNonce::new());
+        let mut pub_nonce = MusigPubNonce(ffi::MusigPubNonce::new());
+        if ffi::secp256k1_musig_nonce_gen(
+            cx,
+            sec_nonce.as_mut_ptr(),
+            pub_nonce.as_mut_ptr(),
+            session_id.as_bytes().as_ptr(),
+            sk_ptr,
+            pub_key.as_c_ptr(),
+            msg_ptr,
+            cache_ptr,
+            extra_ptr,
+        ) == 0
+        {
+            // Rust type system guarantees that
+            // - input secret key is valid
+            // - msg is 32 bytes
+            // - Key agg cache is valid
+            // - extra input is 32 bytes
+            // This can only happen when the session id is all zeros
+            Err(MusigNonceGenError::ZeroSession)
```

Comment:
IMO this should be just `panic`. It can only happen if someone passes wrong value to dangerous ID creation function.